bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/SortedLedgerStorage.java

```diff
@@ -146,14 +165,9 @@ public ByteBuf getEntry(long ledgerId, long entryId) throws IOException {
     }
 
     @Override
-    public Checkpoint checkpoint(final Checkpoint checkpoint) throws IOException {
-        Checkpoint lastCheckpoint = checkpointHolder.getLastCheckpoint();
-        // if checkpoint is less than last checkpoint, we don't need to do checkpoint again.
-        if (lastCheckpoint.compareTo(checkpoint) > 0) {
-            return lastCheckpoint;
-        }
+    public void checkpoint(final Checkpoint checkpoint) throws IOException {
```

Comment:
1) do we really need this method? the actual caller of this method is SortedLedgerStorage.onSizeLimitReached..in that method "memTable.flush(SortedLedgerStorage.this);" is already called after getting 'checkpoint'. I dont see the need of calling memTable.flush(this, checkpoint); again here. 
2) we can remove this method from SortedLedgerStorage class and also we can remove the argument - 'checkpoint' for LedgerStorage.checkpoint method

Reply:
1. we need this method. onSizeLimitReached only flush the memory to entry log files. but it doesn't guarantee data to be flushed to disk. Indeed, it will issue a checkpoint request to checkpointer (which is SyncThread). SyncThread will do the actual checkpoint job. The checkpoint job will ensure the data added before this `checkpoint` will be persisted in entry log files before it update the marker in journal.

2. the argument `checkpoint` is used by memtable to compare if there are data still in memtable that are not flushed to entry log files, when checkpoint starts. it will be used for avoiding unnecessary flush during checkpointing.

Reply:
1) @sijie you misunderstood me. I mean to say do we need to have this overridden method in SortedLedgerStorage? Yes, ofcourse for persistence InterleavedLedgerStorage's checkpoint should be called and I'm not questioning that. My question is why do we need to call "memTable.flush(this, checkpoint);" again? memtable.flush is already called in onSizeLimitReached method, which is the caller for this method.

2) If you agree with my previous point, then you would get why I'm saying this 'checkpoint' argument is not needed for checkpoint method and hence we can remove the argument all together from the method declaration in LedgerStorage interface.

Reply:
@reddycharan - gotcha. I think my 2nd point at previous comment answered your question here :)

the `checkpoint` argument is used by the memtable to compare if there are data is need to flush to entry log files and make it durably stored on entry log files. it is a safeguard here to ensure we got things right. Also `Checkpoint.MAX` can be used for triggering fully flush and checkpoint. Hope this explain your comment.

Reply:
hmm..you still didn't explain me why we need to call memtable.flush(this,checkpoint) here. I've couple of issues with this

1) SortLS.onSizeLimitReached -> SyncThread.checkpoint -> SortLS.checkpoint is the only call hierarchy in production code for this method. In SLS.onSizeLimitReached "memTable.flush(SortedLedgerStorage.this);" is already called, so what is the point in calling it again?

2) Ok, lets assume there is relevant data in EntryMemTable for this checkpoint, which is added after the last memtable.flush call (line 205) (actually that's not the case), now if we call "memTable.flush(this, checkpoint);" here, it would flush the data from memtable to current active log, but the "super.checkpoint(checkpoint);" will flush only rotatedlogs but not active current log, which defies your assumption/purpose

In summary, there is no point in calling "memTable.flush(this, checkpoint);" here and there is no point in having 'checkpoint' argument to the checkpoint method in LedgerStorage interface itself.

Reply:
@reddycharan nice catch here. I have been looking back to all the historic changes we made towards SortedLedgerStorage. I think `checkpoint` (originally known as logMark) as added as a parameter for optimizing skiplist flush, was originally introduced in twitter/boookkeeper#159be8dcb2079e30291aa877f3436126f0579a0b . the point 2 you are raised here is correct for current checkpoint logic. 

So for this logic here, I am going to stick to having a `checkpoint` as the argument to be a safeguard and address the issue at your 2nd point, to be attributing to original design. I will push a change for it.